### PR TITLE
Update Ukkonen for v3

### PR DESCRIPTION
**Pull request by Joseph-Edwards**
This PR updates `ukkonen.hpp` for v3

### CODE REVIEW

**Comment by Joseph-Edwards** [comment]
This is a draft until python bindings are done.

File: include/libsemigroups/ukkonen.hpp [hunk 1]
```diff
@@ -878,343 +1003,664 @@ namespace libsemigroups {
...
+    }
+
+    //! \brief \copybrief Ukkonen.add_word_no_checks
+    //!
+    //! See \ref Ukkonen.add_word_no_checks.
+    //!
+    //! \throws LibsemigroupsException if `throw_if_not_unique_letters(w)`
+    //! throws.
+    //!
+    //! \sa \ref Ukkonen.throw_if_not_unique_letters.
```
Comment (james-d-mitchell) [inline include/libsemigroups/ukkonen.hpp:1078]:
```suggestion
    //! \sa \ref Ukkonen::throw_if_not_unique_letters.
```
Some python slipped back in here

File: include/libsemigroups/ukkonen.hpp [hunk 2]
```diff
@@ -878,343 +1003,664 @@ namespace libsemigroups {
...
+    template <typename Word>
+    void add_word(Ukkonen& u, Word const& w) {
+      add_word(u, w.cbegin(), w.cend());
+    }
+
+    //! \brief \copybrief Ukkonen.add_word_no_checks
```
Comment (james-d-mitchell) [inline include/libsemigroups/ukkonen.hpp:1084]: More python

File: include/libsemigroups/ukkonen.hpp [hunk 3]
```diff
@@ -878,343 +1003,664 @@ namespace libsemigroups {
...
     //!
-    //! \param u the Ukkonen object
-    //! \param words the words to add
+    //! See \ref Ukkonen.add_word_no_checks.
     //!
-    //! \returns None
+    //! \throws LibsemigroupsException if `u.throw_if_not_unique_letters(w, w +
+    //! std::strlen(w))` throws.
+    //!
+    //! \sa \ref Ukkonen.throw_if_not_unique_letters.
```
Comment (james-d-mitchell) [inline include/libsemigroups/ukkonen.hpp:1091]: more python

File: include/libsemigroups/ukkonen.hpp [hunk 4]
```diff
@@ -16,11 +16,6 @@
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 //
 
-// This file contains an implementation of a generalised suffix tree, adapted
```
Comment (james-d-mitchell) [inline include/libsemigroups/ukkonen.hpp:19]: Note sure why these lines are deleted, can you please undelete them?

**Review by james-d-mitchell — CHANGES_REQUESTED** [review]
Looks good apart from the deletion of the lines at the start of the file (which should remain for the attribution if nothing else), and the fact that some python seems to have slipped into the references. I spotted some of these but there might be more, can you please check and fix them all? Thanks!

**Comment by Joseph-Edwards** [comment]
> Looks good apart from the deletion of the lines at the start of the file (which should remain for the attribution if nothing else), and the fact that some python seems to have slipped into the references. I spotted some of these but there might be more, can you please check and fix them all? Thanks!

All done now. The 'cpplint' job will (hopefully) pass when #615 is merged.

With regard to the deleted comments at the top of the file, I had moved those comments into the doxygen comments a bit further down so it's visible in the doc, but have now added it back as a normal comment at the top just to be doubly safe.

**Comment by james-d-mitchell** [comment]
Thanks @Joseph-Edwards can you please rebase onto v3, so that the changes you made for cpplint are picked up?

**Comment by james-d-mitchell** [comment]
Great! Thanks @Joseph-Edwards

**Review by james-d-mitchell — APPROVED** [review]

**Pull request merged**